### ISSUE TITLE
Checklist for Release

### ISSUE THREAD

**Issue by soumith** [pytorch/pytorch#5]
### Core Framework code
- [x] optim + trainer + dataset objects
- [x] Sharing CPU tensors
- [x] Add all operations to autograd
- [x] Free GIL when processing big tensors
- [x] Custom CUDA memory allocator (Sam Gross)
- [x] multi-GPU functions
- [x] nccl integration
- [x] finish legacy.nn
- [x] refactor C API for extensions
- [x] create an example extension with TH/pytorch C API
- [x] Checkpointing and improving torch.save / torch.load to use the same byte order
- [x] implement keyword arguments in cwrap
- [x] go over TH and try to make error messages more descriptive (e.g. if the sizes don't match )
- [ ] Sparse Tensors on CPU and GPU (Zeming)
- [x] improve tensor printing
- [x] functional API for autograd variables
- [x] Finish multiple CUDA types (Soumith)
- [x] Add stochastic nodes
- [x] Add all modules to nn
- [ ] Improved error messages #39 
- [ ] sync legacy.nn with Lua nn
- [ ] move Trainer to torch.experimental
### Operations
- [x] Integrate CuDNN
- [x] Write nn.LSTM*, nn.GRU* etc. to integrate CuDNN RNNs
- [x] Rewrite LookupTable and SparseLinear to use SparseTensors
## FBCode stuff
- [x] Import into FBCode
## Open Source stuff
- [x] Binary builds
- [x] Continuous builds for CUDA
- [ ] MNIST and ResNet18 Contbuilds
- [x] pip wheels
### Lua Bridge
- [ ] integrate lutorpy into pytorch (either as optional package or by default)
   - [ ] change TH indexing to 0-based and add to cwrap the 1-subtraction and addition
### Model Loading
- [x] Build a simple model loader based on https://github.com/bshillingford/python-torchfile
- [x] Apply backward-compatibility patches (they've been removed from legacy nn)
## Framework Integration
- [ ] Caffe2 Integration
  - [ ] Modify TH / THC  / THNN / THCUNN to integrate them
  - [ ] Have a converter that takes in a (Module and input) or (output) and auto-converts it to caffe model
    - [ ] vice versa. take a caffe protobuf and codegen a python class with loading weights
- [ ] Keras Integration
  - [ ] Have a keras backend. Send in a Pull Request to fchollet/keras
- [ ] Converting models between TF and Pytorch
  - [ ] Torch2TF: Same as caffe convertor pretty much!
  - [ ] TF2Torch: same as caffe, but cover ops like tf.if and tf.while
## Website
- [x] Find someone to design and code it
- [x] Getting Started
  - [x] Binary installs
    - [x] anaconda-based which links automatically with MKL
    - [x] Each of them for different CUDA versions. 7.0, 7.5, 8.0
  - [x] Source-based installs
- [ ] Showcase Demos / Examples / ModelZoo elegantly
- [x] Tutorials
- [x] Look at gym.openai.com (http://gym.openai.com/) 
- [x] Developer docs
### Demos / Examples / ModelZoo
- [x] Pre-trained models for each demo (in the model zoo)
  - [x] Create a python wrapper that allows to search and download models (like nltk)
- [x] Simple API for retraining / using pre-trained models on custom datasets
- [x] Documentation on how to modify the example for one's own experiments
- [x] Most or all of them should be multi-GPU ready
### Demos + Examples
- [x] Basic
  - [x] Autoencoders https://github.com/Kaixhin/Autoencoders
- [ ] Vision
  - [ ] Supervised
    - [x] fb.resnet.torch / googlenet for image classification (sam)
    - [x] fastrcnn (francisco)
    - [ ] Video Classification
    - [ ] NeuralTalk2 (paszke)
    - [x] Visual Q&A (paszke)
  - [ ] Unsupervised
    - [x] Image super-resolution (wafi2x) (soumith)
    - [x] DCGANs + Improved Training for GANs + InfoGAN
    - [ ] Text 2 Image (soumith)
    - [ ] Pixel RNNs (soumith)
    - [x] Variational AutoEncoders (joost)
- [x] Games / RL (ludc)
  - [x] Good integration with gym.openai.com (http://gym.openai.com/)
  - [x] RL examples
- [ ] NLP / Text
    - [x] OpenNMT - attention based Seq2seq with MT(Lerer)
  - [ ] Sentiment Analysis
  - [ ] Memory networks for Q & A (memn2n) https://github.com/taey16/MemN2N-python
- [ ] Metalearning
  - [ ] Neural Turing Machine
  - [x] Learning to Learn by Gradient Descent by Gradient Descent
  - [ ] Decoupled Neural Interfaces using Synthetic Gradients https://arxiv.org/abs/1608.05343
- [ ] ConvNet-Benchmarks / DeepMark scripts
### Tutorials
- [x] See #288 
### Documentation
- [x] Auto-generate from source / docstrings
## Links
- [1] http://elm-lang.org/blog/compiler-errors-for-humans
- [2] http://elm-lang.org/blog/compilers-as-assistants

Postponed for next release
-  lazy forward execution engine
-  double backprop
-  Sharing CUDA tensors
 -  look into Cython
-  a built-in profiler for forward/backward (with automatic hints for speeding up the execution?)
## AIViz Integration
-  Have an intial attempt, and talk to Allan

-  Serveable via some Python REST API
  -  figure out details for images and videos

-  Audio
  -  wav2letter
  -  DeepSpeech2 for maybe Switchboard or something (Ask Gabriel)
-  Sparse Models (ads?)

## Distributed Training
-  simple distributed trainer like torch-distlearn / torch-ipc / torch-thrift
-  Synchronous, asynchronous and Elastic SGD
-  Integrate with Andrew / Yangqing's MPI library when that's ready
-  Port image classification and seq2seq to this

- error handling
  -  create a dict for translating exceptions and adding some pytorch specific info, sort of like in Elm [1,2]
  -  make sure there's a clear error message when multiprocessing runs out of fds

**Comment by pltrdy** [comment]
@soumith I see that  Neural Turing Machine was on the list, anything planned about it atm?

**Comment by soumith** [comment]
no, nothing planned.

**Comment by jekbradbury** [comment]
there's a dynamic neural computer implementation that may or may not be complete/fully functional but looks like a good guide to how to implement NTM-like models https://github.com/ypxie/pytorch-NeuCom

**Comment by pltrdy** [comment]
@jekbradbury thx for the link
